File: src/main/java/br/com/sicredi/api/exception/handler/GlobalExceptionHandler.java
package br.com.sicredi.api.exception.handler;

import br.com.sicredi.api.exception.*;
import br.com.sicredi.api.exception.error.ErrorResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.time.Instant;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<ErrorResponse> sendResourceNotFound(ResourceNotFoundException e, HttpServletRequest request) {
        return buildResponseErrorEntity(HttpStatus.NOT_FOUND, e, request, "Resource not found");
    }

    @ExceptionHandler(DataBaseException.class)
    public ResponseEntity<ErrorResponse> sendDataBaseError(DataBaseException e, HttpServletRequest request) {
        return buildResponseErrorEntity(HttpStatus.INTERNAL_SERVER_ERROR, e, request, "Database error");
    }

    @ExceptionHandler(RulingNotFoundException.class)
    public ResponseEntity<ErrorResponse> sendRulingNotFound(RulingNotFoundException e, HttpServletRequest request) {
        return buildResponseErrorEntity(HttpStatus.BAD_REQUEST, e, request, "Ruling not found");
    }

    @ExceptionHandler(SessionAlreadyStartedException.class)
    public ResponseEntity<ErrorResponse> sessionIsStarted(SessionAlreadyStartedException e, HttpServletRequest request) {
        return buildResponseErrorEntity(HttpStatus.CONFLICT, e, request, "Session has already started");
    }

    @ExceptionHandler(SessionClosedException.class)
    public ResponseEntity<ErrorResponse> sessionClosed(SessionClosedException e, HttpServletRequest request) {
        return buildResponseErrorEntity(HttpStatus.CONFLICT, e, request, "Session is closed");
    }

    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ErrorResponse> sendConstraintException(HttpServletRequest request, ConstraintViolationException e) {
        return buildResponseErrorEntity(HttpStatus.INTERNAL_SERVER_ERROR, e, request, "Some attribute is invalid");
    }

    @ExceptionHandler(InvalidCpfException.class)
    public ResponseEntity<ErrorResponse> sendInvalidCpfException(HttpServletRequest request, InvalidCpfException e) {
        return buildResponseErrorEntity(HttpStatus.BAD_REQUEST, e, request, "Invalid CPF");
    }

    @ExceptionHandler(DuplicateVoteException.class)
    public ResponseEntity<ErrorResponse> sendDuplicatedVote(HttpServletRequest request, DuplicateVoteException e) {
        return buildResponseErrorEntity(HttpStatus.CONFLICT, e, request, "The vote is duplicated");
    }

    @ExceptionHandler(InvalidRulingAttributeException.class)
    public ResponseEntity<ErrorResponse> invalidateRulingAttribute(HttpServletRequest request, InvalidRulingAttributeException e) {
        return buildResponseErrorEntity(HttpStatus.BAD_REQUEST, e, request, "Attribute value is not valid");
    }

    private ResponseEntity<ErrorResponse> buildResponseErrorEntity(
            HttpStatus httpStatus, Exception exception,
            HttpServletRequest request, String messageError) {

        log.error(exception.getMessage(), exception);
        var responseError = new ErrorResponse(Instant.now(), httpStatus.value(), messageError, exception.getMessage(), request.getRequestURI());
        return ResponseEntity.status(httpStatus).body(responseError);
    }
}

File: validate-random-cpf/src/main/kotlin/br/com/sicredi/generate_cpf/exception/handler/GlobalExceptionHandler.kt
package br.com.sicredi.generate_cpf.exception.handler

import br.com.sicredi.generate_cpf.exception.ErrorResponse
import br.com.sicredi.generate_cpf.exception.InvalidCpfException
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.client.HttpClientErrorException.NotFound
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.NoHandlerFoundException
import java.time.Instant


@RestControllerAdvice
class GlobalExceptionHandler {

    private val logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ResponseStatus(HttpStatus.NOT_FOUND)
    @ExceptionHandler(InvalidCpfException::class)
    fun sendInvalidCpf(exception: InvalidCpfException, httpServletRequest: HttpServletRequest): ErrorResponse {
        return this.buildResponseErrorEntity(
            HttpStatus.NOT_FOUND,
            exception,
            httpServletRequest,
            "Invalid CPF"
        )
    }

    private fun buildResponseErrorEntity(
        httpStatus: HttpStatus,
        exception: Exception?,
        request: HttpServletRequest,
        messageError: String
    ): ErrorResponse {
        return ErrorResponse(
            Instant.now(),
            httpStatus.value(),
            messageError,
            exception?.message.orEmpty(),
            request.requestURI
        ).also {
            logger.error(messageError)
        }
    }
}